File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """

        stack, node = [], head

        while node:
            stack.append(node)
            node = node.next

        
        start, cnt = head, 0

        while cnt < len(stack) - 1:
            temp = start.next
            node = stack.pop()

            start.next = node
            node.next = temp

            start = temp

            cnt += 1
        
        start.next = None
        

File: 0143-reorder-list/0143-reorder-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
    stack := make([]*ListNode, 0)

    node := head

    for node != nil {
        stack = append(stack, node)
        node = node.Next
    }

    start := head

    cnt := 0
    for cnt < len(stack) - 1 {
        temp := start.Next
        node := stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        start.Next = node
        node.Next = temp

        start = temp

        cnt++
    }

    start.Next = nil
}